Source/Scene/UrlTemplateImageryProvider.js

```diff
@@ -429,12 +440,72 @@ define([
          */
         hasAlphaChannel : {
             get : function() {
+                //>>includeStart('debug', pragmas.debug);
+                if (!this.ready) {
+                    throw new DeveloperError('hasAlphaChannel must not be called before the imagery provider is ready.');
+                }
+                //>>includeEnd('debug');
                 return this._hasAlphaChannel;
             }
         }
     });
 
     /**
+     * Reinitializes this instance.  This method is primarily intended for use
+     * with the `deferReadiness` flag to the constructor.  Reinitializing an instance already in use is supported, but it is not
+     * recommended because existing tiles provided by the imagery provider will not be updated.
+     * @param {Object} options Any of the options that may be passed to the {@see UrlTemplateImageryProvider} constructor.
+     */
+    UrlTemplateImageryProvider.prototype.reinitialize = function(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        this._url = options.url;
```

Comment:
I _think_ that the JavaScript engine will notice that `reinitialize` is called from the constructor function and that it should generate a hidden class for `UrlTemplateImageryProvider` to avoid entering dictionary mode for instances of this class.

However, I would not count on it.  I think we are better off defining all members in the constructor function before calling `reinitialize` and setting them to `undefined`, e.g., 
```javascript
this._url = undefined;
this._pickFeaturesUrl  = undefined;
// ...
this.reinitialize(options);
```